build-support/bin/generate_github_workflows.py

```diff
@@ -0,0 +1,327 @@
+        {
+            "name": "Get commit message for PR builds",
+            "if": "github.event_name == 'pull_request'",
+            "run": 'echo "COMMIT_MESSAGE<<EOF" >> $GITHUB_ENV\necho "$(git log --format=%B -n 1 HEAD^2)" >> $GITHUB_ENV\necho "EOF" >> $GITHUB_ENV\n',
+        },
+    ]
+
+
+def pants_virtualenv_cache() -> Sequence[Step]:
+    return [
+        {
+            "name": "Cache Pants Virtualenv",
+            "uses": "actions/cache@v2",
+            "with": {
+                "path": "~/.cache/pants/pants_dev_deps\n",
+                "key": "${{ runner.os }}-pants-venv-${{ matrix.python-version }}-${{ hashFiles('pants/3rdparty/python/**', 'pants.toml') }}\n",
+            },
+        }
+    ]
+
+
+def pants_config_files() -> Sequence[Step]:
+    return [
+        {
+            "name": "Set pants config files",
+            "run": 'echo \'PANTS_CONFIG_FILES=+["${{ github.workspace }}/pants.ci.toml", "${{ github.workspace }}/pants.remote-cache.toml"]\' >> ${GITHUB_ENV}\n',
+        },
+    ]
+
+
+def bootstrap_caches() -> Sequence[Step]:
+    return [
+        {
+            "name": "Cache Rust toolchain",
+            "uses": "actions/cache@v2",
+            "with": {
+                "path": "~/.rustup/toolchains/${{ env.rust_version }}-*\n~/.rustup/update-hashes\n~/.rustup/settings.toml\n",
+                "key": "${{ runner.os }}-rustup-${{ hashFiles('rust-toolchain') }}",
+            },
+        },
+        {
+            "name": "Cache Cargo",
+            "uses": "actions/cache@v2",
+            "with": {
+                "path": "~/.cargo/registry\n~/.cargo/git\n",
+                "key": "${{ runner.os }}-cargo-${{ hashFiles('rust-toolchain') }}-${{ hashFiles('src/rust/engine/Cargo.*') }}\n",
+                "restore-keys": "${{ runner.os }}-cargo-${{ hashFiles('rust-toolchain') }}-\n",
+            },
+        },
+        *pants_virtualenv_cache(),
+        {
+            "name": "Get Engine Hash",
+            "id": "get-engine-hash",
+            "run": 'echo "::set-output name=hash::$(./build-support/bin/rust/print_engine_hash.sh)"\n',
+            "shell": "bash",
+        },
+        {
+            "name": "Cache Native Engine",
+            "uses": "actions/cache@v2",
+            "with": {
+                "path": "src/python/pants/engine/internals/native_engine.so\nsrc/python/pants/engine/internals/native_engine.so.metadata\n",
+                "key": "${{ runner.os }}-engine-${{ steps.get-engine-hash.outputs.hash }}\n",
+            },
+        },
+    ]
+
+
+def native_engine_so_upload() -> Sequence[Step]:
+    return [
+        {
+            "name": "Upload native_engine.so",
+            "uses": "actions/upload-artifact@v2",
+            "with": {
+                "name": "native_engine.so.${{ matrix.python-version }}.${{ runner.os }}",
+                "path": "src/python/pants/engine/internals/native_engine.so\nsrc/python/pants/engine/internals/native_engine.so.metadata\n",
```

Comment:
These concatenated lines are how `yaml.load(..)` rendered them. This is an MVP from deep within a yak shave, so I'd like to hold off on reformatting these.